File: 572-SubtreeOfAnotherTree/572-SubtreeOfAnotherTree.cpp
// Last updated: 17/07/2025, 20:51:59
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
        bool isSameTree(TreeNode* p, TreeNode* q) {
        if((p == NULL && q != NULL) || (p != NULL && q == NULL)){
            return false;
        }
        if(p == NULL && q == NULL){
            return true;
        }
        if(p->val != q->val){
            return false;
        }
        
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(root == NULL){
            return false;
        }
        bool res = isSameTree(root,subRoot);
        if(res){
            return true;
        }
        return isSubtree(root->left , subRoot) || isSubtree(root->right, subRoot);
    }
};

File: 572-SubtreeOfAnotherTree/572-SubtreeOfAnotherTree.py
# Last updated: 01/10/2025, 00:12:45
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def same(a: Optional[TreeNode], b:Optional[TreeNode]) -> bool:
            if not a and not b:
                return True
            
            if not a or not b  or a.val!=b.val:
                return False
            
            return same(a.left, b.left) and same(a.right, b.right)
        
        if not subRoot:
            return True
        
        if not root:
            return False
        
        if same(root, subRoot):
            return True
        
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        

        